File: app/src/main/java/tomaszkopacz/meetbam/presenters/AccountFriendsFragmentPresenter.java
package tomaszkopacz.meetbam.presenters;

import android.support.v7.widget.RecyclerView;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import tomaszkopacz.meetbam.activities.AccountFriendsFragment;
import tomaszkopacz.meetbam.model.Post;
import tomaszkopacz.meetbam.model.User;
import tomaszkopacz.meetbam.service.LoginService;
import tomaszkopacz.meetbam.service.UserAdapter;
import tomaszkopacz.meetbam.service.UserViewHolder;
import tomaszkopacz.meetbam.service.WebService;

/**
 * Created by tomas on 02.03.2018.
 */

public class AccountFriendsFragmentPresenter implements RecyclerViewPresenter{

    private AccountFriendsFragment fragment;
    private WebService mWebService;
    private LoginService mLoginService;

    // list of users
    private UserAdapter adapter;
    private List<User> users = new ArrayList<>();

    /**
     * Constructor.
     */
    public AccountFriendsFragmentPresenter (AccountFriendsFragment fragment, WebService service){
        this.fragment = fragment;
        this.mWebService = service;
        this.mLoginService = new LoginService(fragment.getActivity().getApplicationContext());
        this.adapter = new UserAdapter(this);
    }

    /**
     * Downloads posts and send them to activity.
     */
    public void downloadPostsList(){

        // remove previous elements
        users.clear();

        // when new elements downloaded
        Call call = mWebService.getFriends(mLoginService.getUserMail());
        call.enqueue(new Callback<List<User>>() {

            @Override
            public void onResponse(Call<List<User>> call, Response<List<User>> response) {

                // assign response objects to list
                users = response.body();

                // send users to activity
                fragment.putUsers(adapter);
            }

            @Override
            public void onFailure(Call<List<User>> call, Throwable t) {

            }
        });
    }

    @Override
    public void onItemBoundAtPosition(RecyclerView.ViewHolder holder, int position) {
        User user = users.get(position);
        ((UserViewHolder)holder).getUsername().setText(user.getName() + " " + user.getSurname());
    }

    @Override
    public int getItemCount() {
        return users.size();
    }
}

File: app/src/main/java/tomaszkopacz/meetbam/presenter/AccountFriendsFragmentPresenter.kt
package tomaszkopacz.meetbam.presenter

import android.support.v7.widget.RecyclerView
import android.view.View
import kotlinx.android.synthetic.main.user_item.view.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import tomaszkopacz.meetbam.R
import tomaszkopacz.meetbam.entity.User
import tomaszkopacz.meetbam.interactor.WebService
import tomaszkopacz.meetbam.view.AccountFriendsFragment
import tomaszkopacz.meetbam.view.MainApp
import tomaszkopacz.meetbam.view.UserAdapter
import java.util.*
import javax.inject.Inject

class AccountFriendsFragmentPresenter(private val fragment: AccountFriendsFragment)
    : RecyclerViewPresenter {

    //service
    val app = fragment.activity!!.application as MainApp
    @Inject lateinit var webService: WebService

    //friends
    private val adapter = UserAdapter(this)
    private var users: MutableList<User>? = ArrayList()

    init {
        (fragment.activity!!.application as MainApp)
                .component!!.inject(this)
    }

    fun downloadPostsList() {

        users!!.clear()

        val call = webService.getFriends(app.getLoggedUser().mail!!)
        call.enqueue(object : Callback<List<User>> {

            override fun onResponse(call: Call<List<User>>, response: Response<List<User>>) {
                users = response.body() as MutableList<User>?
                fragment.putUsers(adapter)
            }

            override fun onFailure(call: Call<List<User>>, t: Throwable) {

            }
        })
    }

    override fun onItemBoundAtPosition(holder: RecyclerView.ViewHolder, position: Int) {
        val user = users!![position]

        holder as UserAdapter.UserViewHolder
        holder.itemView.username.text = fragment.getString(R.string.custom_user, user.name, user.surname)
    }

    override fun onItemClick(view: View) {

    }

    override fun onLongItemClick(view: View) {

    }

    override fun itemCount(): Int {
        return users!!.size
    }
}
